samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,109 @@
-  public interface VoidFunction3<A, B, C> {
-    public void apply(A a, B b, C c);
-  }
-
-  /**
-   * Method to apply a map function (1:1) on a {@link MessageStream}
-   *
-   * @param mapper  the mapper function to map one input {@link Message} to one output {@link Message}
+   * @param mapFn  the function to map one input {@link Message} to its corresponding output {@link Message}
    * @param <OM>  the type of the output {@link Message} in the output {@link MessageStream}
-   * @return the output {@link MessageStream} by applying the map function on the input {@link MessageStream}
+   * @return the output {@link MessageStream} after applying the map function on the input {@link MessageStream}
    */
-  public <OM extends Message> MessageStream<OM> map(Function<M, OM> mapper) {
-    Operator<OM> op = Operators.<M, OM>getStreamOperator(m -> new ArrayList<OM>() { {
-        OM r = mapper.apply(m);
-        if (r != null) {
-          this.add(r);
-        }
-      } });
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  <OM extends Message> MessageStream<OM> map(MapFunction<M, OM> mapFn);
 
   /**
-   * Method to apply a flatMap function (1:n) on a {@link MessageStream}
+   * Applies the 1:n {@link FlatMapFunction} to {@link Message} in this {@link MessageStream}, concatenates the results,
+   * and returns the resulting {@link MessageStream}.
+   * <p>
+   * For example, consider an input {@link MessageStream} containing 2 {@link Message}s. If the flatMap function
+   * produces 3 transformed {@link Message}s for each input {@link Message}, the resulting {@link MessageStream}
+   * contains the 6 transformed {@link Message}s.
    *
-   * @param flatMapper  the flat mapper function to map one input {@link Message} to zero or more output {@link Message}s
+   * @param flatMapFn  the flatMap function to map one input {@link Message} to zero or more output {@link Message}
    * @param <OM>  the type of the output {@link Message} in the output {@link MessageStream}
-   * @return the output {@link MessageStream} by applying the map function on the input {@link MessageStream}
+   * @return the output {@link MessageStream} after applying the flatMap function on the input {@link MessageStream}
    */
-  public <OM extends Message> MessageStream<OM> flatMap(Function<M, Collection<OM>> flatMapper) {
-    Operator<OM> op = Operators.getStreamOperator(flatMapper);
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  <OM extends Message> MessageStream<OM> flatMap(FlatMapFunction<M, OM> flatMapFn);
 
   /**
-   * Method to apply a filter function on a {@link MessageStream}
+   * Applies the {@link FilterFunction} to this {@link MessageStream} and returns the resulting {@link MessageStream}.
    *
-   * @param filter  the filter function to filter input {@link Message}s from the input {@link MessageStream}
+   * @param filterFn  the filter function to filter {@link Message}s from the input {@link MessageStream}
```

Comment:
Thanks @prateekm! 

Also, cloud dataflow defines the following characteristics for all their functions: (map, flatMap, ParDo)
1. Idempotency
2. Thread-compatibility
3. Serializability

https://cloud.google.com/dataflow/model/transforms#UserCodeReqs